DOCSP-36217: polymorphic serialization (#163)
mongodb/docs-kotlin: source/fundamentals/data-formats/serialization.txt

- `BsonConfiguration <{+api+}/apidocs/bson-kotlinx/bson-kotlinx/org.bson.codecs.kotlinx/-bson-configuration/index.html>`__
 
.. _kotlin-polymorphic:

Polymorphic Serialization
-------------------------

The {+driver-short+} natively supports serialization and deserialization
of polymorphic classes. When you mark a sealed interface and data
classes that inherit that interface with the ``@Serializable``
annotation, the driver uses a ``KSerializer`` implementation to handle
conversion of your types to and from BSON.

When you insert an instance of a polymorphic data class into MongoDB,
the driver adds the field ``_t``, the
discriminator field. The value of this field is the data class name.

Polymorphic Data Classes Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In this example, you create an interface and two data classes that
inherit that interface. In the data classes, the ``id`` field is marked
with the annotations described in the
:ref:`kotlin-data-class-annotation` section:

.. literalinclude:: /examples/generated/KotlinXSerializationTest.snippet.polymorphic-dataclasses.kt
    :language: kotlin

> Then, you can perform operations with data classes as usual. In the
> following example, you parametrize the collection with the ``Person``
> interface, but you can perform operations with the polymorphic classes
> ``Teacher`` and ``Student``. When you retrieve documents, the driver
> automatically detects the type based on the discriminator value and
> deserializes them accordingly.

Reviewer: Nit (these are based on the way we seem to typically introduce examples, let me know if I'm on the wrong track here).

```suggestion
Then, you can perform operations with data classes as usual. The following
example parametrizes the collection with the ``Person``
interface and performs operations with the polymorphic classes
``Teacher`` and ``Student``. When you retrieve documents, the driver
automatically detects the type based on the discriminator value and
deserializes them accordingly.
```
